vshard/router/init.lua

```diff
@@ -873,79 +1022,165 @@ local function router_map_callrw(router, func, args, opts)
+            group_count = group_count + 1
+            table.insert(replicasets_to_check, replicasets_all[uuid])
+        end
+        timeout, err, err_id = wait_connected_to_masters(
+            replicasets_to_check, timeout)
+        if not timeout then
+            goto fail
+        end
+
+        -- Send ref requests with timeouts to the replicasets.
+        futures = table_new(0, group_count)
+        for id, buckets in pairs(grouped_buckets) do
+            if timeout == nil then
                 err_id = id
                 goto fail
             end
-            -- Map returns true,res or nil,err.
-            res = res:iterator()
-            local count = res:decode_array_header()
-            ok = res:decode()
-            if ok == nil then
-                err = res:decode()
+            local args_ref
+            if replicasets_to_map[id] then
+                -- Replicaset is already referenced on a previous iteration.
+                -- Simply get the moved buckets without double referencing.
+                args_ref = {'storage_moved_buckets', buckets}
+            else
+                args_ref = {'storage_ref_with_lookup', rid, timeout, buckets}
+            end
+            res, err = replicasets_all[id]:callrw('vshard.storage._call',
+                                                  args_ref, opts_async)
+            if res == nil then
                 err_id = id
                 goto fail
             end
-            if count > 1 then
-                map[id] = res:take_array(count - 1)
-            end
-            timeout = deadline - fiber_clock()
+            futures[id] = res
         end
-    else
+
+        -- Wait for the refs to be done and collect moved buckets.
+        bucket_ids = {}
         for id, f in pairs(futures) do
             res, err = future_wait(f, timeout)
+            -- Handle netbox error first.
             if res == nil then
                 err_id = id
                 goto fail
             end
-            -- Map returns true,res or nil,err.
-            ok, res = res[1], res[2]
-            if ok == nil then
-                err = res
+            -- Ref returns nil,err or {rid, moved}.
+            res, err = res[1], res[2]
+            if res == nil then
                 err_id = id
                 goto fail
             end
-            if res ~= nil then
-                -- Store as a table so in future it could be extended for
-                -- multireturn.
-                map[id] = {res}
+            for _, bucket_id in pairs(res.moved) do
+                bucket_reset(router, bucket_id)
+                table.insert(bucket_ids, bucket_id)
+            end
+            if res.rid then
+                assert(not replicasets_to_map[id])
+                -- If there are no buckets on the replicaset, it would not be
+                -- referenced.
+                replicasets_to_map[id] = replicasets_all[id]
             end
             timeout = deadline - fiber_clock()
```

Comment:
We can early exit from the loop `while #bucket_ids > 0` here, if `#replicasets_to_map == #replicasets_all`. We anyway referenced all masters, no need to continue searching for the buckets, we don't care.

E.g. when we have two replicasets:
1 - buckets 1, 2, 6;
2 - buckets 4, 5, 3.
They exchanged buckets 6 and 3 with each other, router doesn't know about this yet. If `bucket_ids` is {1, 4, 3} we don't have to know the location of bucket 3 to continue.

Or maybe I'm missing smth?

Reply:
Thanks, about this I was thinking for a while after you noticed it. The problem is that `#` is not safe on non-array types. It returns invalid untrue result. Which means we can't check equality of 2 maps `replicasets_to_map` and `replicasets_all` for O(1). We would need to count `replicasets_all` for a linear time and maintain a counter for `replicasets_to_map`. Even when there is no rebalancing, on the common path. It is possible, but makes it hard to keep this counter in sync for `replicasets_all`, because this counter would need to be a member of the router.

I couldn't find another easy way how to optimize it, so I've left it as it. If you have any other simple ideas, then I can apply!

Reply:
> (which I don't really get why)

Because it increases the number of states to maintain. We need to keep replicaset count and replicaset dictionary in sync, while they are separate objects. My primary goal (after not having bugs) is to keep the code clean and simple, and only then performance (unless degradation is so severe, that it is a bug). In this case I would say that bringing this counter would increase the cognitive weight of the code, while not giving a significant boost or fixing a bug. Does that make sense?

Reply:
> Does that make sense?

Yeap)